### ISSUE TITLE
How to show the address of ingress

### ISSUE THREAD

**Issue by junyeong-huray** [haproxytech/kubernetes-ingress#480]
Hello

I have two ingress controllers in one k8s cluster.

one thing is `ingress-nginx` and the other is `haproxy-kubernetes-ingress`.

I use `ingress-nginx` to handle requests from the internet and it is bound to
a cloud provisioned loadbalancer which is assigned a public IP.
And I use `haproxy-kubernetes-ingress` to handle requests from the internal 
backend servers outside of the k8s cluster or from the browsers of developers
in the office. And the ingress is bound to  a cloud provisioned loadbalancer
which is assigned a private IP.

The thing is that the result of command, `kubectl get ing`, does not show the address
of ingress which is handled by `haproxy-kubernetes-ingress`..

Here is the output of kubectl (the actual domain name is masked by XXXX):

```command
$ kubectl get ing
NAME               CLASS    HOSTS                     ADDRESS                               PORTS     AGE
chase-private      <none>   chase.XXXXX                                                     80        35m
chase-public       <none>   chase.XXXXX     XXXXX.kr.lb.naverncp.com   80, 443   23m
```

The ingress controller of `chase-private` is `haproxy-kubernetes-ingress` and
the ingress controller of `chase-public` is `ingress-nginx`.

The output of `kubectl get ing/chase-private -o yaml` is here:

```yaml
$ kubectl get ing/chase-private -o yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    haproxy.org/ingress.class: haproxy
  name: chase-private
    ... omitted ...
spec:
  rules:
  - host: chase.XXXXX
    http:
      paths:
      - backend:
          service:
            name: chase
            port:
              number: 8000
        path: /
        pathType: Prefix
status:
  loadBalancer: {}
```

Look at that `loadBalancer` is empty.

But the output of `kubectl get ing/chase-public -o yaml` shows `loadBalancer` as expected:

```yaml
apiVersion: networking.k8s.io/v1
kind: Ingress
metadata:
  annotations:
    kubernetes.io/ingress.class: nginx
    nginx.ingress.kubernetes.io/auth-url: https://$host/oauth2/auth
    nginx.ingress.kubernetes.io/backend-protocol: HTTP
    nginx.ingress.kubernetes.io/force-ssl-redirect: "true"
  name: chase-public
  ... omitted ...
spec:
  rules:
  - host: chase.XXXXX
    http:
      paths:
      - backend:
          service:
            name: chase
            port:
              number: 8000
        path: /
        pathType: Prefix
  tls:
  - hosts:
     ... omitted ...
status:
  loadBalancer:
    ingress:
    - hostname: XXXXXXXXXXX.kr.lb.naverncp.com
```

What can I do to make `chase-private` have the address of `loadBalancer`?
Any advice will be appreciated.

Thanks,

**Comment by junyeong-huray** [comment]
And here is the k8s manifest that is used to install haproxy-kubernetes-ingress.
It is based on the [official manifest](https://raw.githubusercontent.com/haproxytech/kubernetes-ingress/v1.8/deploy/haproxy-ingress.yaml) but is modified little bit.

```yaml
---
apiVersion: v1
kind: Namespace
metadata:
  name: haproxy-controller

---
apiVersion: v1
kind: ServiceAccount
metadata:
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller

---
kind: ClusterRole
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: haproxy-kubernetes-ingress
rules:
- apiGroups:
  - ""
  resources:
  - configmaps
  - endpoints
  - nodes
  - pods
  - services
  - namespaces
  - events
  - serviceaccounts
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - "extensions"
  - "networking.k8s.io"
  resources:
  - ingresses
  - ingresses/status
  - ingressclasses
  verbs:
  - get
  - list
  - watch
- apiGroups:
  - "extensions"
  - "networking.k8s.io"
  resources:
  - ingresses/status
  verbs:
  - update
- apiGroups:
  - ""
  resources:
  - secrets
  verbs:
  - get
  - list
  - watch
  - create
  - patch
  - update
- apiGroups:
  - core.haproxy.org
  resources:
  - '*'
  verbs:
  - get
  - list
  - watch
  - update
- apiGroups:
  - "discovery.k8s.io"
  resources:
  - endpointslices
  verbs:
  - get
  - list
  - watch
---
kind: ClusterRoleBinding
apiVersion: rbac.authorization.k8s.io/v1
metadata:
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller
roleRef:
  apiGroup: rbac.authorization.k8s.io
  kind: ClusterRole
  name: haproxy-kubernetes-ingress
subjects:
- kind: ServiceAccount
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller

---
apiVersion: v1
kind: ConfigMap
metadata:
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller
data:

---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    run: ingress-default-backend
  name: haproxy-kubernetes-ingress-default-backend
  namespace: haproxy-controller
spec:
  replicas: 1
  selector:
    matchLabels:
      run: ingress-default-backend
  template:
    metadata:
      labels:
        run: ingress-default-backend
    spec:
      containers:
      - name: ingress-default-backend
        image: gcr.io/google_containers/defaultbackend:1.0
        ports:
        - containerPort: 8080

---
apiVersion: v1
kind: Service
metadata:
  labels:
    run: ingress-default-backend
  name: haproxy-kubernetes-ingress-default-backend
  namespace: haproxy-controller
spec:
  selector:
    run: ingress-default-backend
  ports:
  - name: port-1
    port: 8080
    protocol: TCP
    targetPort: 8080

---
apiVersion: apps/v1
kind: Deployment
metadata:
  labels:
    run: haproxy-ingress
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller
spec:
  replicas: 1
  selector:
    matchLabels:
      run: haproxy-ingress
  template:
    metadata:
      labels:
        run: haproxy-ingress
    spec:
      serviceAccountName: haproxy-kubernetes-ingress
      containers:
      - name: haproxy-ingress
        image: haproxytech/kubernetes-ingress:1.8.4
        args:
          - --configmap=haproxy-controller/haproxy-kubernetes-ingress
          - --default-backend-service=haproxy-controller/haproxy-kubernetes-ingress-default-backend
          - --ingress.class=haproxy
          - --disable-ipv6
        securityContext:
          runAsUser:  1000
          runAsGroup: 1000
          capabilities:
            drop:
              - ALL
            add:
              - NET_BIND_SERVICE
        resources:
          requests:
            cpu: "500m"
            memory: "50Mi"
        livenessProbe:
          httpGet:
            path: /healthz
            port: 1042
        ports:
        - name: http
          containerPort: 80
        - name: https
          containerPort: 443
        - name: stat
          containerPort: 1024
        env:
        - name: TZ
          value: "Etc/UTC"
        - name: POD_NAME
          valueFrom:
            fieldRef:
              fieldPath: metadata.name
        - name: POD_NAMESPACE
          valueFrom:
            fieldRef:
              fieldPath: metadata.namespace
---
apiVersion: v1
kind: Service
metadata:
  labels:
    run: haproxy-ingress
  name: haproxy-kubernetes-ingress
  namespace: haproxy-controller
  annotations:
    # this configuration must be set to make the cloud LB private
    service.beta.kubernetes.io/ncloud-load-balancer-internal: "true"
spec:
  selector:
    run: haproxy-ingress
  type: LoadBalancer
  ports:
  - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  - name: https
    port: 443
    protocol: TCP
    targetPort: 443
  - name: stat
    port: 1024
    protocol: TCP
    targetPort: 1024

``` 

The difference is here:

```diff
--- haproxy-ingress.yaml.orig   2022-08-24 18:53:39.000000000 +0900
+++ haproxy-ingress.yaml        2022-08-19 18:27:07.000000000 +0900
@@ -163,10 +163,12 @@
       serviceAccountName: haproxy-kubernetes-ingress
       containers:
       - name: haproxy-ingress
-        image: haproxytech/kubernetes-ingress
+        image: haproxytech/kubernetes-ingress:1.8.4
         args:
           - --configmap=haproxy-controller/haproxy-kubernetes-ingress
           - --default-backend-service=haproxy-controller/haproxy-kubernetes-ingress-default-backend
+          - --ingress.class=haproxy
+          - --disable-ipv6
         securityContext:
           runAsUser:  1000
           runAsGroup: 1000
@@ -209,10 +211,13 @@
     run: haproxy-ingress
   name: haproxy-kubernetes-ingress
   namespace: haproxy-controller
+  annotations:
+    # this configuration must be set to make the cloud LB private
+    service.beta.kubernetes.io/ncloud-load-balancer-internal: "true"
 spec:
   selector:
     run: haproxy-ingress
-  type: NodePort
+  type: LoadBalancer
   ports:
   - name: http
     port: 80
```

**Comment by ivanmatmati** [comment]
Hi, You should use [publish service](https://www.haproxy.com/documentation/kubernetes/latest/configuration/controller/#--publish-service).

**Comment by junyeong-huray** [comment]
@ivanmatmati The emptiness of `status.loadBalancer.ingress` in ingress resource makes argocd misread the health of ingress.
https://argo-cd.readthedocs.io/en/stable/faq/#why-is-my-application-stuck-in-progressing-state

So I need to solve this situation and it might be great if haproxy-kubernetes-ingress shows status of loadbalancer in a ingress resource correctly.

**Comment by junyeong-huray** [comment]
@ivanmatmati
Oh thanks, I'll try your suggestion.

**Comment by junyeong-huray** [comment]
@ivanmatmati 
I just found that the IC pod is repeating restarting. It seems that the pod exited abnormally repeatedly. I'll figure it out.

**Comment by junyeong-huray** [comment]
@ivanmatmati 
The log of the IC pod shows lots of this error message
```
expected argument for flag `--publish-service'
```

it seems that I have to provide argument to `--publish-service`, for example, `--publish-service XXXXXXXXX.kr.lb.naverncp.com`

Is it right?


I didn't do this kind of job when I installed `ingress-nginx`, I wonder how `ingress-nginx` publishes the address of LB to ingress resources automatically.
And I hope `haproxy-kubernetes-ingress` also do this automatically. Is it possible with `haproxy-kubernetes-ingress:1.8.4` now?

**Comment by junyeong-huray** [comment]
Although I added --publish-service=XXXXXXXXX to the IC pod and recreated the ingress resource,
the resource does not contain the status of loadbalancer.

I was excited when I got to know about --publish-service since I thought it would be gonna resolve my problem quickly. But nothing had been changed and I still have the same problem. What a disappointment.

Anyway is there anything I can try to solve this problem?
Thanks,

**Comment by junyeong-huray** [comment]
@ivanmatmati 
Your solution works eventually.
It was my fault. I misunderstood the reference.

At first, I thought only `--publish-service` is needed without additional argument for it.
And then I got to know `--publish-service` needs an argument but I misunderstood that it takes an address of load balancer.
Finally I realized that I have to give `<namespace>/<service>` to `--publish-service`. What a fool.
I apologize that I misunderstood your suggestion.

And I am satisfied with this solution. It works for me now.
